### PR TITLE
Fix pg dataloader

### PR DESCRIPTION
**Pull request by sunghyuckhong**
## 문제상황
- MLPerf submission model로 full model & full calibration 수행시 OOM 발생
- v3.11에서 paged_attenttion calib dataloader를 model forward전에 GPU에 항상 옮겨놨었는데, 이로 인하여 calibration for loop에서 gpu에 total block space 2개가 복사되는 시점이 발생해서 OOM이 발생함

## 목적(해결방향)
- num_blocks를 줄임

## 구현 설명 Abstract
-

## 구현 설명 Detail (ex. 추가된 argument)
- Dataloader_for_paged_attention에서는 모든 dataset이 하나의 kv_cache block을 공유하기 떄문에, n번째 calibration forward시 생성된 KV_cache block이 n+1번째 calibratio forward시 영향을 줄 수 있음
- 이 현상을 방지하기 위하여 empty_cache 정의 

## test 통과 내역 (CI 통과내역과 어떤 machine (GPU pod or NPU machien or local)에서 테스트했는지)
- [ ] local machine with 3090
- [ ] GPU pod
- [ ] warboy pod
  - `"python main.py --scenario Offline --model-path ./model/ --dataset-path ./data/cnn_eval_accuracy_ci.json --model_script_path ./quantization/model_script/Qlevel4_RGDA0-W8A8KV8-PTQ-SMQ.yaml --gpu --accuracy --use_mcp --calib-dataset-path ./data/cnn_dailymail_calibration.json --recalibrate --model_source [transformers or furiosa_llm_original]"

## TODO (ex. 후속PR예고)
-

### CODE REVIEW

File: language/gpt-j/quantization/get_quant_model.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def get_total_block_space(config, batch_size = 1, block_size = 1, bucket_size =
     else:
         raise NotImplementedError
     
-    num_blocks = batch_size * 4  * 2 * (bucket_size) * block_size + 1
+    num_blocks = batch_size * 2 * (bucket_size) * block_size + 1
```
Comment (jeongin-yun) [inline language/gpt-j/quantization/get_quant_model.py]: 왜 block 갯수를 줄여도 되는건가요?

File: language/gpt-j/quantization/calibration_utils/dataset_paged_attention.py [hunk 2]
```diff
@@ -35,5 +35,14 @@ def move_common_tensors_to_device(self, device, device_map = None):
...
+    def empty_total_block_space(self, device, device_map = None):
+        tensor_shape = self.total_block_space[0][0].shape
+        dtype = self.total_block_space[0][0].dtype
+        for type_idx in range(len(self.total_block_space)):
+            for block_idx in range(len(self.total_block_space[type_idx])):
+                kv_cache_device = device
+                if device_map is not None and len(device_map) > 0:
+                    if isinstance(device[str(type_idx)], int):
+                        kv_cache_device = device[str(type_idx)]
+                self.total_block_space[type_idx][block_idx] = torch.zeros(tensor_shape, dtype = dtype).to(kv_cache_device)
```
Comment (jeongin-yun) [inline language/gpt-j/quantization/calibration_utils/dataset_paged_attention.py]: 이건 calibraiton 시 self.total_block_space 를 initialize 해주기 위한거죠?

**Comment by jeongin-yun** [comment]
gptj mlperf_submission model calibration 를 위한 반영이니, 우선은 test 가능한 성혁님 branch 에 merge 하는게 좋을거 같습니다.

File: language/gpt-j/quantization/get_quant_model.py (hunk 1 above)
Comment (sunghyuckhong) [reply language/gpt-j/quantization/get_quant_model.py]: 구두로 설명드리긴 했는데, 이 block은 KV_cache를 저장하는 공간이라 qparam에 영향을 주지 못하기 때문입니다. 참고로 generator에 쓰는 num_block과 이는 별개입니다. (+혹시라도 generate에 영향을 주는지 확인하기 위하여 branch에서 test진행하겠습니다)

File: language/gpt-j/quantization/calibration_utils/dataset_paged_attention.py (hunk 2 above)
Comment (sunghyuckhong) [reply language/gpt-j/quantization/calibration_utils/dataset_paged_attention.py]: 네 맞습니다. 구현 설명 Detail에서 설명드렸던 현상을 해소하기 위함입니다.

**Comment by jeongin-yun** [comment]
> gptj mlperf_submission model calibration 를 위한 반영이니, 우선은 test 가능한 성혁님 branch 에 merge 하는게 좋을거 같습니다.

정정 할게요. block size 변경이 evaluation 시 generate에 영향을 주는지 여부만 확인되면 main 에 merge 하면 될거 같습니다

**Comment by sunghyuckhong** [comment]
> > gptj mlperf_submission model calibration 를 위한 반영이니, 우선은 test 가능한 성혁님 branch 에 merge 하는게 좋을거 같습니다.
> 
> 정정 할게요. block size 변경이 evaluation 시 generate에 영향을 주는지 여부만 확인되면 main 에 merge 하면 될거 같습니다
 
RunTimeError 발생 안하는 것 확인했습니다.  https://furiosa-ai.slack.com/archives/C05DK2P1LH0/p1719913076919839?thread_ts=1719842234.292219&cid=C05DK2P1LH0

**Comment by jeongin-yun** [comment]
결국 torch.zeros 로 값만 초기화 하니, empty_total_block_space 보단 init_total_block_space 가 더 직관적인거 같은데, 어떻게 생각하시나요?

**Comment by sunghyuckhong** [comment]
넵 method명 수정해 놓고, slack에서 말씀드렸던 num_block 줄이는 commit도 곧 올리겠습니다.

**Comment by sunghyuckhong** [comment]
> 결국 torch.zeros 로 값만 초기화 하니, empty_total_block_space 보단 init_total_block_space 가 더 직관적인거 같은데, 어떻게 생각하시나요?

넵 method명 수정해 놓고, slack에서 말씀드렸던 num_block 줄이는 commit도 곧 올리겠습니다.

**Comment by sunghyuckhong** [comment]
@jeongin-yun 양쪽 PR에 반영해놨으니, 확인부탁드립니다.

**Review by jeongin-yun — APPROVED** [review]
LGTM
